config/initializers/diaspora_federation.rb

```diff
@@ -104,6 +104,17 @@
 
     on :receive_entity do |entity, sender, recipient_id|
       Person.by_account_identifier(sender).pod.try(:schedule_check_if_needed)
+      if User.exists?(recipient_id)
+        User.find(recipient_id).tap do |user|
```

Comment:
If you use `User.find_by(id: recipient_id)` it returns `nil` if the user doesn't exist. That way you can skip the extra exists query, since all users should exist at this point the exists is kinda redundant anyway.

Reply:
@SuperTux88 This runs into many test-errors because user.person line 108 seems to be nil.
Is this because of the tests (the build invalid user entities) or is a person not present in any case?

Reply:
You also removed the `&` on the `next unless` line, they are needed there because they may be nil there. But if they are nil, it already exits there, so the additional `&` at the line below is never needed because at that point it's already ensured that the user and the person exists.

But now thinking about that, it might also only be an issue with the test and incomplete test data, and in the real world both the user and the person should (in theory) always be existing ... but it's a bit strange that so many tests fail. But it probably doesn't hurt to keep them in that line, but it isn't needed a few lines below when it was already handled here.

Reply:
I think I know what the problem is. `recipient_id` is sometimes `nil` (well is always `nil` for public messages). So checking if that's nil, and only do the whole thing when a message is sent to a user directly (and this user was migrated). But for public federation where `recipient_id` is `nil` we don't even need to do the `User.find_by(id: recipient_id)`.

Reply:
So `recipient_id` comes from here (if it isn't `nil`):
https://github.com/diaspora/diaspora/blob/16e754f4c71f1389a4fd6c5ef5e9a778d8285aa7/config/initializers/diaspora_federation.rb#L101

And if `recipient_id` isn't `nil`, then neither `user` nor `user.person` should ever be nil. So the only check that is needed is if `recipient_id` is `nil`, if it isn't, no `&` should be needed anymore.